### PR TITLE
Add data_interfaces

### PR DESCRIPTION
**Pull request by welpaolo**
This PR adds the data_interfaces library in the Kafka K8s charm.

### CODE REVIEW

File: tests/integration/test_provider.py [hunk 1]
```diff
@@ -197,14 +146,109 @@ async def test_remove_application_removes_user_and_acls(ops_test: OpsTest, usern
...
+    for acl in load_acls(model_full_name=ops_test.model_full_name, zookeeper_uri=zookeeper_uri):
+        assert acl.username in usernames
+        assert acl.operation in ["READ", "DESCRIBE", "CREATE", "WRITE"]
+        assert acl.resource_type in ["GROUP", "TOPIC"]
+        if acl.resource_type == "TOPIC":
+            assert acl.resource_name == "test-topic"
+
+    # remove application
+    # await ops_test.model.remove_application(DUMMY_NAME_1, block_until_done=True)
+    # HERE
```
Comment (marcoppenheimer) [inline tests/integration/test_provider.py:183]: **nitpick:** Comments can be removed I think.

File: tests/integration/test_provider.py [hunk 2]
```diff
@@ -197,14 +146,109 @@ async def test_remove_application_removes_user_and_acls(ops_test: OpsTest, usern
...
+    super_users = load_super_users(model_full_name=ops_test.model_full_name)
+    assert len(super_users) == 1
+
+    await ops_test.build_charm("tests/integration/app-charm")
+
+    # await asyncio.gather(
+    #     ops_test.model.deploy(
+    #         app_charm, application_name=DUMMY_NAME_1, num_units=1, series="focal"
+    #     )
+    # )
```
Comment (marcoppenheimer) [inline tests/integration/test_provider.py:203]: **nitpick:** Comments can be removed here.

File: tests/integration/test_provider.py [hunk 3]
```diff
@@ -197,14 +146,109 @@ async def test_remove_application_removes_user_and_acls(ops_test: OpsTest, usern
...
+    assert ops_test.model.applications[APP_NAME].status == "active"
+
+    super_users = load_super_users(model_full_name=ops_test.model_full_name)
+    assert len(super_users) == 1
+
+
+@pytest.mark.abort_on_fail
+async def test_connection_updated_on_tls_enabled(ops_test: OpsTest, app_charm: PosixPath):
+    """Test relation when TLS is enabled."""
+    # await ops_test.model.deploy(app_charm, application_name=DUMMY_NAME_1, num_units=1),
```
Comment (marcoppenheimer) [inline tests/integration/test_provider.py:236]: **nitpick:** Same here.

**Review by marcoppenheimer — CHANGES_REQUESTED** [review]
Left some minor comments and nitpicks, but overall I think this is really solid and a massive improvement on the code quality of the charm. 

Really happy you chose to tidy up bits beyond the scope of the ticket, it's made a big difference.

File: src/provider.py [hunk 4]
```diff
@@ -32,134 +27,78 @@ def __init__(self, charm) -> None:
...
+        self.kafka_provider.set_zookeeper_uris(relation.id, zookeeper_uris)
 
-        # non-leader units need cluster_config_changed event to update their super.users
-        self.charm.model.get_relation(PEER).data[self.charm.app].update(
-            {provider_relation_config["username"]: provider_relation_config["password"]}
-        )
+    @property
+    def peer_relation(self) -> Relation:
+        """The Kafka cluster's peer relation."""
+        return self.charm.model.get_relation(PEER)
```
[the remark this answers is not in the document]
Comment (deusebio) [reply src/provider.py:101]: I guess this is only reshuffling position of a method that was already there, but a bit further up in the file. I'm not sure what your point @marcoppenheimer

**Review by deusebio — DISMISSED** [review]
Very good work @welpaolo !

File: tests/integration/test_charm.py [hunk 5]
```diff
@@ -55,22 +56,15 @@ async def test_logs_write_to_storage(ops_test: OpsTest):
     await asyncio.gather(
         ops_test.model.deploy(app_charm, application_name=DUMMY_NAME, num_units=1, series="jammy"),
     )
-    await ops_test.model.wait_for_idle(apps=[APP_NAME, DUMMY_NAME], timeout=1000)
-    await ops_test.model.add_relation(APP_NAME, DUMMY_NAME)
+    await ops_test.model.wait_for_idle(apps=[APP_NAME, DUMMY_NAME])
+    await ops_test.model.add_relation(APP_NAME, f"{DUMMY_NAME}:{REL_NAME_ADMIN}")
```
[the remark this answers is not in the document]
Comment (welpaolo) [reply tests/integration/test_charm.py:60]: Yes this is right!

**Review by marcoppenheimer — APPROVED** [review]
Ayy tests pass, nice job! LGTM

**Review by deusebio — APPROVED** [review]
All good!

**Pull request merged**